[quant] added dq->op->q quantization patterns for GELU, bmm and softmax ops

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56004 [quant] added dq->op->q quantization patterns for GELU and softmax ops**

Summary: added reference pattern support for GELU, softmax and bmm for int dtypes. For GELU and Softmax, this consisted of adding reference patterns to the default node handler for int dtypes. Note GELU and softmax patterns are not registered since they do not have a proper quantized kernel which means they would either add unnecessary dequant and quant ops to the network, or they would simply error. This can be circumvented with custom qconfig usage as in test_gelu_reference

bmm was added within binary ops along with some significant changes to how that code is structured. Theoretically the reference pattern used for bmm could be applied to other dtypes. This was not enabled because of issues relating to Line 1323 in quantize.py. In essence, the prepare step does not know whether an op will use a reference pattern or not, so for ops that are supported with one dtype in reference and one dtype normally, this has the potential to cause issues. This is difficult to get aorund with the is_reference flag being available in the prepare step or discussed changes around separating

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_gelu_reference
python test/test_quantization.py TestQuantizeFxOps.ttest_gelu_normal
python test/test_quantization.py TestQuantizeFxOps.test_softmax_reference
python test/test_quantization.py TestQuantizeFxOps.test_softmax_normal
python test/test_quantization.py TestQuantizeFxOps.test_silu_reference
python test/test_quantization.py TestQuantizeFxOps.test_bmm_int_reference
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestFuseFx
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxModels

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27818340](https://our.internmc.facebook.com/intern/diff/D27818340)

HDCharles committed:
added dq->op->q quantization patterns for GELU and softmax ops

Summary: Used the defaultNodeQuantizationHandler to add ops, not sure how to test the int functionality, the int tensors don't like being on cpu. I edited the silu test in order to share code

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_gelu
python test/test_quantization.py TestQuantizeFxOps.test_softmax
python test/test_quantization.py TestQuantizeFxOps.test_silu

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -814,11 +818,15 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
             torch.nn.InstanceNorm3d: int8_dtypes,
             torch.nn.LayerNorm: all_dtypes,
             torch.nn.SiLU: fp16_dtypes,
+            torch.nn.GELU: all_dtypes,
+            torch.nn.Softmax: all_dtypes,
```
should this be `fp16_dtypes`?

jerryzh168 replied:
are we adding this for int8 or fp16? I saw int8 in the docs

HDCharles replied:
I thoguht the goal was for int8 compatibility. From the design doc:

"The way we can implement this in FX graph mode quantization is to add a "reference pattern" for gelu, which mimics gelu_int8 without an actual gelu_int8 kernel."

vkuzo replied:
@HDCharles , you're right.  It should be `int8` for the current request, I misspoke earlier.  `all_dtypes` sounds good.

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -814,11 +818,15 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
             torch.nn.functional.layer_norm: all_dtypes,
             torch.nn.functional.leaky_relu: int8_dtypes,
             torch.nn.functional.silu: fp16_dtypes,
+            torch.nn.functional.gelu: all_dtypes,
+            torch.nn.functional.softmax: all_dtypes,
```
and this

vkuzo commented on test/quantization/test_quantize_fx.py:
```diff
@@ -3066,29 +3066,44 @@ def test_instance_norm(self):
+            M(module, functional), (data.float(),), quant_type, custom_qconfig_dict=qconfig_dict,
             expected_node_occurrence=node_occurrence)
 
+    def test_gelu(self):
```
can we also add a test that with a qconfig for `int8`, we get regular floating point gelu/softmax/silu

HDCharles replied:
I tried to add that, but kept getting errors about ints not having a cpu kernel. This comment seems contradictory to your comment above about dtypes, I think I am misunderstanding the difference between int8 in the qconfig vs in the dtype, I thought the one depended on support of the other, is that not the case?

jerryzh168 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -3066,29 +3066,44 @@ def test_instance_norm(self):
                 return x
 
-        data = (torch.randn((2, 2, 2, 2), dtype=torch.float),)
+        data = torch.randn((2, 2, 2, 2), dtype=torch.float16)
```
this should be float I think, we'll convert it to fp16 in the model.

jerryzh168 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -3066,29 +3066,44 @@ def test_instance_norm(self):
+        m = M(module, functional).eval()
+        m_prep = torch.quantization.quantize_fx.prepare_fx(m, qconfig_dict)
+        m_quant = torch.quantization.quantize_fx.convert_fx(m_prep)
+        self.assertEqual(m_quant(data),m(data.float()), rtol=1e-3, atol=1e-4)
```
we don't really need to test this I think, there is not much guarantee for numerics for float vs quantized.

HDCharles replied:
perhaps I should add a test, but the intention was to make sure that the model runs correctly with float16 data (the other testing piece doesn't do that), but it sounds like from your other comment that we don't need that?

jerryzh168 replied:
yeah running the model sounds OK. the other tests just makes sure we generate the correct quantized op

jerryzh168 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -3066,29 +3066,44 @@ def test_instance_norm(self):
+        m_quant = torch.quantization.quantize_fx.convert_fx(m_prep)
+        self.assertEqual(m_quant(data),m(data.float()), rtol=1e-3, atol=1e-4)
+
         node_occurrence = {
```
we can test for dequant - fp op - quant patterns here to make sure the patterns are correct with expected_node_list, example: https://github.com/pytorch/pytorch/blob/master/test/quantization/test_quantize_fx.py#L891

HDCharles replied:
ok, will do

HDCharles committed:
Update on "[quant] added dq->op->q quantization patterns for GELU and softmax ops"


Summary: Used the defaultNodeQuantizationHandler to add ops, not sure how to test the int functionality, the int tensors don't like being on cpu. I edited the silu test in order to share code

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_gelu
python test/test_quantization.py TestQuantizeFxOps.test_softmax
python test/test_quantization.py TestQuantizeFxOps.test_silu

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "[quant] added dq->op->q quantization patterns for GELU and softmax ops"


Summary: Used the defaultNodeQuantizationHandler to add ops, not sure how to test the int functionality, the int tensors don't like being on cpu. I edited the silu test in order to share code

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_gelu
python test/test_quantization.py TestQuantizeFxOps.test_softmax
python test/test_quantization.py TestQuantizeFxOps.test_silu

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -767,14 +767,19 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
 @register_quant_pattern(torch.nn.InstanceNorm3d)
 @register_quant_pattern(torch.nn.LayerNorm)
 @register_quant_pattern(torch.nn.SiLU)
+# @register_quant_pattern(torch.nn.GELU)
+# @register_quant_pattern(torch.nn.Softmax)
```
can we add a comment that this is commented out because we only support reference patterns for these and not a real kernel?

HDCharles replied:
yeah, done

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -767,14 +767,19 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
 @register_quant_pattern(torch.nn.functional.layer_norm)
 @register_quant_pattern(torch.nn.functional.leaky_relu)
 @register_quant_pattern(torch.nn.functional.silu)
+# @register_quant_pattern(torch.nn.functional.gelu)
+# @register_quant_pattern(torch.nn.functional.softmax)
```
same here

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -830,7 +839,7 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
 
         # TODO: make helper functions for (torch.quint8, torch.qint8, None)
-        if dtypes in [(torch.quint8, torch.qint8, None)]:
+        if not is_reference and dtypes in [(torch.quint8, torch.qint8, None)]:
```
optional readability nit: might be nice to restructure these if statements like this

```
if dtypes in [(torch.quint8, torch.qint8, None)]:
  if not is_reference:
    ...
  else:  # is_reference
    ...
else:
  assert dtypes in [(torch.float16, torch.float16, None)]
  if not is_reference:
    ...
  else:  # is_reference
    ...
```

this way the order is consistent

HDCharles replied:
I wasn't sure if it would be more readable to explicate the 4 modes or to split them with 2x2 branches. If that is considered more readable I can change it.

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -861,18 +870,37 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
                 return quantizer.quantized_graph.create_node(
                     "call_function", quantized_op, args, kwargs)
+        elif is_reference and dtypes in [(torch.quint8, torch.qint8, None)]:
+            args = load_arg(quantized=[0])(node.args)
```
is this line needed? would it get overriden by the next line

HDCharles replied:
i think its the load_arg statement that is doing something, I'm pulling this from the convrelu reference code so if its redundant then its also redundant there.

vkuzo reviewed (commented):
looks good! had some small comments inline

HDCharles committed:
Update on "[quant] added dq->op->q quantization patterns for GELU and softmax ops"


Summary: Used the defaultNodeQuantizationHandler to add ops, not sure how to test the int functionality, the int tensors don't like being on cpu. I edited the silu test in order to share code

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_gelu_reference
python test/test_quantization.py TestQuantizeFxOps.ttest_gelu_normal
python test/test_quantization.py TestQuantizeFxOps.test_softmax_reference
python test/test_quantization.py TestQuantizeFxOps.test_softmax_normal
python test/test_quantization.py TestQuantizeFxOps.test_silu_reference
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "[quant] added dq->op->q quantization patterns for GELU and softmax ops"


Summary: Used the defaultNodeQuantizationHandler to add ops, not sure how to test the int functionality, the int tensors don't like being on cpu. I edited the silu test in order to share code

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_gelu_reference
python test/test_quantization.py TestQuantizeFxOps.ttest_gelu_normal
python test/test_quantization.py TestQuantizeFxOps.test_softmax_reference
python test/test_quantization.py TestQuantizeFxOps.test_softmax_normal
python test/test_quantization.py TestQuantizeFxOps.test_silu_reference
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "[quant] added dq->op->q quantization patterns for GELU and softmax ops"


Summary: Used the defaultNodeQuantizationHandler to add ops, not sure how to test the int functionality, the int tensors don't like being on cpu. I edited the silu test in order to share code

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_gelu_reference
python test/test_quantization.py TestQuantizeFxOps.ttest_gelu_normal
python test/test_quantization.py TestQuantizeFxOps.test_softmax_reference
python test/test_quantization.py TestQuantizeFxOps.test_softmax_normal
python test/test_quantization.py TestQuantizeFxOps.test_silu_reference
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on test/quantization/test_quantize_fx.py:
```diff
@@ -3092,28 +3094,145 @@ def test_instance_norm(self):
-        m = self.checkGraphModeFxOp(
-            M(), data, quant_type, custom_qconfig_dict=qconfig_dict,
-            expected_node_occurrence=node_occurrence)
+        custom_qconfig_dict = {"additional_quant_pattern": additional_quant_pattern_dict}
```
nit: this is called `prepare_custom_config_dict` in the API definition, which is different from `qconfig_dict`.  Might be worth renaming this var, so it's consistent with the API definition, and this technically isn't a qconfig_dict.

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -170,60 +176,81 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
-                    'call_function', self.quantized_binary_op,
-                    load_arg(quantized=[quantized_index])(self.binary_op_node.args), self.binary_op_node.kwargs)
-            else:
+        if is_reference and self.binary_op in binary_reference_op_supported_dtypes and dtypes in binary_reference_op_supported_dtypes[self.binary_op]:
```
formatting nit: can we do this to keep line length manageable

```
if (
  foo and
  bar
):
  baz
```

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -775,14 +803,23 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
 @register_quant_pattern(torch.nn.LayerNorm)
 @register_quant_pattern(torch.nn.SiLU)
+# we currently only support reference patterns for these ops so they have been removed
+# until they receive a proper fp16 kernal. To use the reference pattern, use a custom qconfig
```
nit: `kernel`

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -775,14 +803,23 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
 @register_quant_pattern(torch.nn.functional.leaky_relu)
 @register_quant_pattern(torch.nn.functional.silu)
+# we currently only support reference patterns for these ops so they have been removed
+# until they receive a proper fp16 kernal. To use the reference pattern, use a custom qconfig
```
nit: `kernel`

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -1317,11 +1318,9 @@ def record_match(pattern, node, matched):
-                                    (dtypes not in binary_op_supported_dtypes[node.target])
-                                )
-
+                                skip_this_match = (node.target in binary_op_supported_dtypes and dtypes not in binary_op_supported_dtypes[node.target]) and not (
```
@jerryzh168 , thoughts about making is_reference passed to prepare so we don't have to do hacks like this?  If is_reference is needed for proper placement of observers, it makes sense for prepare to know about it.

jerryzh168 replied:
not sure what is this guarding against, I saw both binary_op_supported_dtypes and binary_reference_op_supported_dtypes here, does passing is_reference help here?

vkuzo replied:
the original `skip_this_match` was a hack added by me.  Essentially the current matching logic only matches on the op pattern, and does not know about qconfig or is_reference while matching.  But, the actual logic needed is to match on the combination of pattern, qconfig for that pattern, and the value of is_reference.

For example, if `is_reference` was known at prepare time, this check could be written as

```
skip_this_match_if_not_is_reference = (
  (not is_reference) and
  (node.target in binary_op_supported_dtypes) and
  (dtypes not in binary_op_supported_dtypes[node.target])
)

skip_this_match_if_is_reference = (
  is_reference and
  node.target in binary_reference_op_supported_dtypes and 
  dtypes in binary_reference_op_supported_dtypes[node.target]
)

skip_this_match = skip_this_match_if_not_is_reference or skip_this_match_if_is_reference
```

vkuzo replied:
@HDCharles , could we go with something like this for the meantime for the sake of readability, as we probably will need to do the "when shoudl is_reference flag be set" discussion async

```
# note: the value of is_reference is unknown at prepare, so we have to cover both cases
# the first case is when is_reference = False
skip_this_match_if_not_is_reference = ( 
  (node.target in binary_op_supported_dtypes) and 
  (dtypes not in binary_op_supported_dtypes[node.target]) 
) 
# the first case is when is_reference = True
skip_this_match_if_is_reference = ( 
  (node.target in binary_reference_op_supported_dtypes) and 
  (dtypes in binary_reference_op_supported_dtypes[node.target])
) 
skip_this_match = skip_this_match_if_not_is_reference or skip_this_match_if_is_reference
```

HDCharles replied:
do you mean skip if both are true? what you have is going to gate anything not supported on both

vkuzo replied:
sure, I was more commenting on the readability, spelling out what exactly we know and don't know in this function, and clearly explaining what the condition is doing, keeping lines short, etc.

raghuramank100 replied:
For ops like bmm, whether or not we place observers depends on is_reference. We should therefore have prepare take this as an argument too. Can we follow up on this in a later PR?

vkuzo reviewed (commented):
lg, had one question for @jerryzh168

raghuramank100 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -3092,28 +3094,145 @@ def test_instance_norm(self):
+        self._test_default_node_quant_handler_ops(
+            module, functional, qconfig, is_reference, node_list)
+
+    def test_gelu_reference(self):
```
Can we combine the reference tests for gelu, softmax and silu into a single one?

raghuramank100 replied:
Got it, feels like an acceptable tradeoff: larger test files, but cleaner test signals.

raghuramank100 commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -170,60 +176,81 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
+                    zero_point = int(zero_point)
+                    scale_arg, zero_point_arg = create_qparam_nodes(quantizer, node.name, scale, zero_point)
+
+                    if self.relu_node is not None:
```
Why are we hard coding the op to add /add_relu here? Should this be deleted (cc @jerryzh168 )

HDCharles replied:
Yeah it seems redundant but this code is unchanged from before.

jerryzh168 replied:
yeah looks like this can be deleted @HDCharles please delete them

raghuramank100 commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -170,60 +176,81 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
-                    'call_function', self.quantized_binary_op,
-                    load_arg(quantized=[quantized_index])(self.binary_op_node.args), self.binary_op_node.kwargs)
-            else:
+        if is_reference and self.binary_op in binary_reference_op_supported_dtypes and dtypes in binary_reference_op_supported_dtypes[self.binary_op]:
```
I am not following the control logic here, can you update once you fix the linter issues

HDCharles replied:
ok this is updated, to clarify, if we have a data structure like:

dict[op]-> [list of supported dtypes for op]

to see if a dtype/op combo is supported, we have to first check if op is in dict, and then check if dtype is in dict[op].

raghuramank100 commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -775,14 +803,23 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
 @register_quant_pattern(torch.nn.functional.layer_norm)
 @register_quant_pattern(torch.nn.functional.leaky_relu)
 @register_quant_pattern(torch.nn.functional.silu)
+# we currently only support reference patterns for these ops so they have been removed
```
These were not present before, so I am not sure I follow the comment. Do we need it?

HDCharles replied:
the comment was requested https://github.com/pytorch/pytorch/pull/56004/files#r614173173

I think the intention is to clarify why its unregistered here despite it being explicitly supported in our list of ops/dtypes.

raghuramank100 commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -836,50 +877,72 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
         else:
-            assert dtypes == (torch.float16, torch.float16, None)
-            return quantizer.quantized_graph.node_copy(node, load_arg(quantized=False))
+            assert is_reference
```
NIT: Why not check for this explicitly instead of having an else followed by an assert?

HDCharles replied:
I was mostly replicating lines like: https://github.com/pytorch/pytorch/blob/3802e577fbce9cb690d050d09b5288a3c97d5694/torch/quantization/fx/quantization_patterns.py#L225

HDCharles committed:
Update on "[quant] added dq->op->q quantization patterns for GELU and softmax ops"


Summary: added reference pattern support for GELU, softmax and bmm for int dtypes. For GELU and Softmax, this consisted of adding reference patterns to the default node handler for int dtypes. Note GELU and softmax patterns are not registered since they do not have a proper quantized kernel which means they would either add unnecessary dequant and quant ops to the network, or they would simply error. This can be circumvented with custom qconfig usage as in test_gelu_reference

bmm was added within binary ops along with some significant changes to how that code is structured. Theoretically the reference pattern used for bmm could be applied to other dtypes. This was not enabled because of issues relating to Line 1323 in quantize.py. In essence, the prepare step does not know whether an op will use a reference pattern or not, so for ops that are supported with one dtype in reference and one dtype normally, this has the potential to cause issues. This is difficult to get aorund with the is_reference flag being available in the prepare step or discussed changes around separating

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_gelu_reference
python test/test_quantization.py TestQuantizeFxOps.ttest_gelu_normal
python test/test_quantization.py TestQuantizeFxOps.test_softmax_reference
python test/test_quantization.py TestQuantizeFxOps.test_softmax_normal
python test/test_quantization.py TestQuantizeFxOps.test_silu_reference
python test/test_quantization.py TestQuantizeFxOps.test_bmm_int_reference
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestFuseFx
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxModels

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27818340](https://our.internmc.facebook.com/intern/diff/D27818340)

[ghstack-poisoned]

HDCharles committed:
Update on "[quant] added dq->op->q quantization patterns for GELU and softmax ops"


Summary: added reference pattern support for GELU, softmax and bmm for int dtypes. For GELU and Softmax, this consisted of adding reference patterns to the default node handler for int dtypes. Note GELU and softmax patterns are not registered since they do not have a proper quantized kernel which means they would either add unnecessary dequant and quant ops to the network, or they would simply error. This can be circumvented with custom qconfig usage as in test_gelu_reference

bmm was added within binary ops along with some significant changes to how that code is structured. Theoretically the reference pattern used for bmm could be applied to other dtypes. This was not enabled because of issues relating to Line 1323 in quantize.py. In essence, the prepare step does not know whether an op will use a reference pattern or not, so for ops that are supported with one dtype in reference and one dtype normally, this has the potential to cause issues. This is difficult to get aorund with the is_reference flag being available in the prepare step or discussed changes around separating

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_gelu_reference
python test/test_quantization.py TestQuantizeFxOps.ttest_gelu_normal
python test/test_quantization.py TestQuantizeFxOps.test_softmax_reference
python test/test_quantization.py TestQuantizeFxOps.test_softmax_normal
python test/test_quantization.py TestQuantizeFxOps.test_silu_reference
python test/test_quantization.py TestQuantizeFxOps.test_bmm_int_reference
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestFuseFx
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxModels

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27818340](https://our.internmc.facebook.com/intern/diff/D27818340)

[ghstack-poisoned]

HDCharles committed:
Update on "[quant] added dq->op->q quantization patterns for GELU and softmax ops"


Summary: added reference pattern support for GELU, softmax and bmm for int dtypes. For GELU and Softmax, this consisted of adding reference patterns to the default node handler for int dtypes. Note GELU and softmax patterns are not registered since they do not have a proper quantized kernel which means they would either add unnecessary dequant and quant ops to the network, or they would simply error. This can be circumvented with custom qconfig usage as in test_gelu_reference

bmm was added within binary ops along with some significant changes to how that code is structured. Theoretically the reference pattern used for bmm could be applied to other dtypes. This was not enabled because of issues relating to Line 1323 in quantize.py. In essence, the prepare step does not know whether an op will use a reference pattern or not, so for ops that are supported with one dtype in reference and one dtype normally, this has the potential to cause issues. This is difficult to get aorund with the is_reference flag being available in the prepare step or discussed changes around separating

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_gelu_reference
python test/test_quantization.py TestQuantizeFxOps.ttest_gelu_normal
python test/test_quantization.py TestQuantizeFxOps.test_softmax_reference
python test/test_quantization.py TestQuantizeFxOps.test_softmax_normal
python test/test_quantization.py TestQuantizeFxOps.test_silu_reference
python test/test_quantization.py TestQuantizeFxOps.test_bmm_int_reference
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestFuseFx
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxModels

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27818340](https://our.internmc.facebook.com/intern/diff/D27818340)

[ghstack-poisoned]

jerryzh168 reviewed (approved):
lgtm, thanks for addressing all the comments

HDCharles committed:
Update on "[quant] added dq->op->q quantization patterns for GELU and softmax ops"


Summary: added reference pattern support for GELU, softmax and bmm for int dtypes. For GELU and Softmax, this consisted of adding reference patterns to the default node handler for int dtypes. Note GELU and softmax patterns are not registered since they do not have a proper quantized kernel which means they would either add unnecessary dequant and quant ops to the network, or they would simply error. This can be circumvented with custom qconfig usage as in test_gelu_reference

bmm was added within binary ops along with some significant changes to how that code is structured. Theoretically the reference pattern used for bmm could be applied to other dtypes. This was not enabled because of issues relating to Line 1323 in quantize.py. In essence, the prepare step does not know whether an op will use a reference pattern or not, so for ops that are supported with one dtype in reference and one dtype normally, this has the potential to cause issues. This is difficult to get aorund with the is_reference flag being available in the prepare step or discussed changes around separating

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_gelu_reference
python test/test_quantization.py TestQuantizeFxOps.ttest_gelu_normal
python test/test_quantization.py TestQuantizeFxOps.test_softmax_reference
python test/test_quantization.py TestQuantizeFxOps.test_softmax_normal
python test/test_quantization.py TestQuantizeFxOps.test_silu_reference
python test/test_quantization.py TestQuantizeFxOps.test_bmm_int_reference
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestFuseFx
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxModels

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27818340](https://our.internmc.facebook.com/intern/diff/D27818340)

[ghstack-poisoned]

HDCharles committed:
Update on "[quant] added dq->op->q quantization patterns for GELU and softmax ops"


Summary: added reference pattern support for GELU, softmax and bmm for int dtypes. For GELU and Softmax, this consisted of adding reference patterns to the default node handler for int dtypes. Note GELU and softmax patterns are not registered since they do not have a proper quantized kernel which means they would either add unnecessary dequant and quant ops to the network, or they would simply error. This can be circumvented with custom qconfig usage as in test_gelu_reference

bmm was added within binary ops along with some significant changes to how that code is structured. Theoretically the reference pattern used for bmm could be applied to other dtypes. This was not enabled because of issues relating to Line 1323 in quantize.py. In essence, the prepare step does not know whether an op will use a reference pattern or not, so for ops that are supported with one dtype in reference and one dtype normally, this has the potential to cause issues. This is difficult to get aorund with the is_reference flag being available in the prepare step or discussed changes around separating

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_gelu_reference
python test/test_quantization.py TestQuantizeFxOps.ttest_gelu_normal
python test/test_quantization.py TestQuantizeFxOps.test_softmax_reference
python test/test_quantization.py TestQuantizeFxOps.test_softmax_normal
python test/test_quantization.py TestQuantizeFxOps.test_silu_reference
python test/test_quantization.py TestQuantizeFxOps.test_bmm_int_reference
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestFuseFx
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxModels

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27818340](https://our.internmc.facebook.com/intern/diff/D27818340)

[ghstack-poisoned]
